### PR TITLE
Working add shipping class option and modal

### LINKED ISSUE #37886: Product Block Editor: Allow creation of shipping class
**Description**

Users should be able to create a new shipping class within the new UI as an option within the shipping class dropdown.
This should behave similar to our previous implementation (see screenshots of previous implementation).

Image

Image

**Designs**

AkVNGImLgSqCObTQ3idVn7-fi-1005_134968

**Acceptance criteria**

- [ ] `Add new shipping class` option should be present in shipping class dropdown.
- [ ] `New shipping class` modal should show up when `Add new shipping class` is selected.
- [ ] Modal should contain `name`, `slug` (optional), and `description` (optional) fields and an `Add` button that creates the shipping class
- [ ] When a user clicks `Add` the shipping class should be created, the Modal close and the new item to be auto selected within the dropdown.

### PR DESCRIPTION
**Pull request by joelclimbsthings**
### Submission Review Guidelines:

- I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
- I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
- I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/). 
- Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

Adding "create new shipping class" option in Shipping class dropdown on product block editor, as well as modal.

Closes #37886 .

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Checkout branch and enable `product-block-editor` feature flag with the WCA Test Helper plugin.
2. Navigate to Products -> Add new, and you should see the product block editor.
3. Navigate to Shipping tab, and select "Follow class" under "Shipping fee" heading.
4. This should reveal a drop-down menu with a list of any shipping classes.
5. Click "Add new shipping class" within the drop-down, which should reveal the shipping class modal.
6. Fill out the modal form with a shipping class and hit "Add"
7. This should add the shipping class and select it within the drop-down.
8. Save the product, and refresh to ensure the value persists.

### CODE REVIEW

File: packages/js/product-editor/src/blocks/shipping-fee/edit.tsx [hunk 1]
```diff
@@ -23,17 +23,41 @@ import {
...
+
+type ServerErrorResponse = {
+	code: string;
+};
 
 const FOLLOW_CLASS_OPTION_VALUE = 'follow_class';
 const FREE_SHIPPING_OPTION_VALUE = 'free_shipping';
 
+export const DEFAULT_SHIPPING_CLASS_OPTIONS: SelectControl.Option[] = [
+	{ value: '', label: __( 'No shipping class', 'woocommerce' ) },
```
Comment (louwie17) [inline packages/js/product-editor/src/blocks/shipping-fee/edit.tsx:45]: It seems like when we select this now, the `Free Shipping` button gets auto selected. Is this what we decided on?
It feels kinda strange, maybe we want to update that `useEffect` on line 174 to fix this.

**Review by louwie17 — DISMISSED** [review]
Nice work @joelclimbsthings, this tested well! Although I did encounter a strange problem with the `No shipping class` option, where it auto selected the free shipping after selecting that.
I assume this was slightly un-expected behaviour after adding the `No shipping class` option? This could be addressed in a follow up, it seemed like the rest tested fine.

**Comment by joelclimbsthings** [comment]
Thanks for the review @louwie17 . When I saw that behavior I made the assumption that it was expected since it's an invalid state, but on second though it _is_ odd, so I've removed it in https://github.com/woocommerce/woocommerce/pull/37968/commits/68016b814d07d0735776e60fd4687b653ea30c00. 

I also had to rebase, but that's the extent of the changes. Let me know what you think.

File: packages/js/product-editor/src/components/add-new-shipping-class-modal/add-new-shipping-class-modal.tsx [hunk 2]
```diff
@@ -40,15 +38,12 @@ function ShippingClassForm( { onAdd, onCancel }: ShippingClassFormProps ) {
...
-				label={ interpolateComponents( {
-					mixedString: __( 'Name {{required/}}', 'woocommerce' ),
-					components: {
-						required: (
-							<span className="woocommerce-add-new-shipping-class-modal__optional-input">
-								{ __( '(required)', 'woocommerce' ) }
-							</span>
-						),
-					},
+				label={ createInterpolateElement( 'Name <required />', {
```
Comment (joshuatf) [inline packages/js/product-editor/src/components/add-new-shipping-class-modal/add-new-shipping-class-modal.tsx]: I think we'll need to make this string translatable:

```suggestion
				label={ createInterpolateElement( __( 'Name <required />', 'woocommerce' ), {
```

**Review by joshuatf — COMMENTED** [review]
Code looks good to me.  Left one comment about a translatable string and then same question as @louwie17 on the behavior.

File: packages/js/product-editor/src/components/add-new-shipping-class-modal/add-new-shipping-class-modal.tsx (hunk 2 above)
Comment (joelclimbsthings) [reply packages/js/product-editor/src/components/add-new-shipping-class-modal/add-new-shipping-class-modal.tsx]: Nice catch! Done in 442b9746cfa0d1623759860d9131326e2dec810a

**Comment by joelclimbsthings** [comment]
Thanks @joshuatf , I believe I'd already removed that behavior in https://github.com/woocommerce/woocommerce/commit/68016b814d07d0735776e60fd4687b653ea30c00 as mentioned [here](https://github.com/woocommerce/woocommerce/pull/37968#issuecomment-1527851467). Was there something different happening?

**Review by louwie17 — APPROVED** [review]
This tested well this time around, nice work @joelclimbsthings 👍 

I did notice that `No shipping class` is shown as an invalid option, disabling the form:
Screenshot 2023-05-01 at 12 01 48 PM
This is not a blocker for me in this PR, but curious if that was the expected behaviour we intended? cc: @jarekmorawski

**Comment by joelclimbsthings** [comment]
Thanks @louwie17 , I assumed this was correct when "follow class" selected under shipping fee, since it wouldn't make sense if no shipping class is selected. Definitely up to change via a follow-up pending @jarekmorawski 's thoughts!

**Comment by rodelgc** [comment]
Updated testing instructions for this PR:

1. Checkout branch and enable `product-block-editor` feature flag with the WCA Test Helper plugin.
2. Navigate to Products -> Add new, and you should see the product block editor.
3. Navigate to Shipping tab, and click on the "Shipping class" menu.
4. This should reveal a drop-down menu with a list of any shipping classes.
5. Click "Add new shipping class" within the drop-down, which should reveal the shipping class modal.
6. Fill out the modal form with a shipping class and hit "Add"
7. This should add the shipping class and select it within the drop-down.
8. Save the product, and refresh to ensure the value persists.

**Pull request merged**